File: polyglot_repo/go/retry.go
package util

import (
	"math/rand"
	"time"
)

// Retry runs fn up to attempts times, sleeping with exponential backoff
// (plus jitter) between failures, and returns the last error.
func Retry(attempts int, base time.Duration, fn func() error) error {
	var err error
	for i := 0; i < attempts; i++ {
		if err = fn(); err == nil {
			return nil
		}
		sleep := base * time.Duration(1<<i)
		sleep += time.Duration(rand.Int63n(int64(base)))
		time.Sleep(sleep)
	}
	return err
}

// Backoff describes an exponential backoff schedule.
type Backoff struct {
	Base   time.Duration
	Factor float64
}

// Next returns the delay before attempt i (0-based).
func (b Backoff) Next(i int) time.Duration {
	d := float64(b.Base)
	for j := 0; j < i; j++ {
		d *= b.Factor
	}
	return time.Duration(d)
}

File: sample_repo/utils/retry.py
"""Resilience helpers."""
import random
import time


def retry_with_backoff(fn, retries=3, base_delay=0.5, factor=2.0, jitter=0.1,
                       retry_on=(Exception,)):
    """Call `fn` with exponential backoff on failure.

    Retries up to `retries` times, sleeping base_delay * factor**attempt
    (plus uniform jitter) between attempts, then re-raises the last
    error. Use this for any flaky network call instead of hand-rolling
    retry loops.
    """
    attempt = 0
    while True:
        try:
            return fn()
        except retry_on:
            attempt += 1
            if attempt > retries:
                raise
            # exponential backoff with a little jitter to avoid stampedes
            delay = base_delay * (factor ** (attempt - 1))
            delay += random.uniform(0, jitter)
            time.sleep(delay)
